internal/kms/kms.go

```diff
@@ -216,6 +221,344 @@ func (k *Kms) CreateKeys(ctx context.Context, scopeId string, opt ...Option) err
+			ctx, cancel = context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+		}
+		// Update the progress of this run
+		args := []any{
+			sql.Named("key_id", run.KeyId),
+			sql.Named("table_name", tableName),
+		}
+		if _, err := k.writer.Exec(ctx, fmt.Sprintf(updateCompletedCountQueryTemplate, tableName), args); err != nil {
+			if retErr != nil {
+				// Just emit an error, don't mutate the returned error
+				_ = errors.E(ctx, errors.WithWrap(err))
+			} else {
+				retErr = err
+			}
+		}
+	}()
+
+	rows, err = k.reader.Query(ctx, dataKeyVersionIdScopeIdQuery, []any{run.KeyId})
+	if err != nil {
+		return errors.Wrap(ctx, err, op, errors.WithMsg("failed to get scope id for data key version"))
+	}
+	defer rows.Close()
+	var scopeId string
+	for rows.Next() {
+		if err := k.reader.ScanRows(ctx, rows, &scopeId); err != nil {
+			return errors.Wrap(ctx, err, op, errors.WithMsg("failed to scan scope id for data key version"))
+		}
+	}
+
+	// Call the function to rewrap the data in the table. The progress will be automatically
+	// updated by the deferred function.
+	return rewrapFn(ctx, run.KeyId, scopeId, k.reader, k.writer, k)
+}
+
+// MonitorDataKeyVersionDestruction monitors any pending destruction jobs. If
+// a job has finished rewrapping all rows, it will destroy the key version
+// by calling RevokeKeyVersion on the underlying kms.
+func (k *Kms) MonitorDataKeyVersionDestruction(ctx context.Context) error {
+	const op = "kms.(Kms).MonitorDataKeyVersionDestruction"
+
+	rows, err := k.reader.Query(ctx, finishedDestructionJobsQuery, nil)
+	if err != nil {
+		return errors.Wrap(ctx, err, op, errors.WithMsg("failed to find completed destruction jobs"))
+	}
+	defer rows.Close()
+	var completedDataKeyVersionIds []string
+	for rows.Next() {
+		if err := k.reader.ScanRows(ctx, rows, &completedDataKeyVersionIds); err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+	}
+	for _, dataKeyVersionId := range completedDataKeyVersionIds {
+		// Finally, revoke the key, deleting it from the database.
+		// This will error if anything still references it that isn't
+		// meant to be cascade deleted.
+		if err := k.underlying.RevokeKeyVersion(ctx, dataKeyVersionId); err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+	}
+	return nil
+}
+
+// DestroyKeyVersion starts the process of destroying a key version in the scope.
+// If the key version is a KEK, this simply rewraps any existing DEK version's encrypted with
+// the KEK and destroys the key version. If the key version is a DEK, this creates a
+// key version destruction job, which will rewrap existing data and destroy the key
+// asynchronously. As a special case, if the DEK currently encrypts no data, the DEK is
+// immediately destroyed. The boolean return value indicates whether the key was immediately
+// destroyed.
+// Options are ignored.
+func (k *Kms) DestroyKeyVersion(ctx context.Context, scopeId string, keyVersionId string, _ ...Option) (bool, error) {
```

Comment:
do we want to check for missing parameters?

Reply:
Added checks and tests for this